### ISSUE TITLE
no-unstable-nested-components gives false-positive error if nested component is "calculated" inside a function.

### ISSUE THREAD

**Issue by avasuro** [jsx-eslint/eslint-plugin-react#3636]
### Is there an existing issue for this?

- [X] I have searched the existing issues and my issue is unique
- [X] My issue appears in the command-line and not only in the text editor

### Description Overview

<details open>
  <summary>Click to show code example</summary>
  
  ```javascript
  
const Test = ({ componentProp }) => {
    return <div>{componentProp}</div>
}

const call = fn => fn()

export const App = () => {
    const someCondition: number = 1

    const componentInVariable = (() => {
        switch (someCondition) {
            case 1:
                return <div>one</div>
            case 2:
                return <div>two</div>
            default:
                return <div>many</div>
        }
    })() // <-- note that this is a self-invoked function, so it immediately returns ready-to-use jsx element

    return (
        <>
            {/* Test case 1: */}
            <Test
                componentProp={
                    (() => {
                        switch (someCondition) {
                            case 1:
                                return <div>one</div>
                            case 2:
                                return <div>two</div>
                            default:
                                return <div>many</div>
                        }
                    })()  // <-- note that this is a self-invoked function, so it immediately returns ready-to-use jsx element
                }
            />
            
            {/* Test case 2: */}
            <Test
                // Note that `call` immediately runs given function, so ready-to-use jsx element is passed into `componentProp`:
                componentProp={call(() => {
                    switch (someCondition) {
                        case 1:
                            return <div>one</div>
                        case 2:
                            return <div>two</div>
                        default:
                            return <div>many</div>
                    }
                })}
            />

            {/* Test case 3: */}
            <Test componentProp={componentInVariable} />
        </>
    )
}
  ```
</details>

### Expected Behavior

For the example above:

**Actual result:**

- ❌ Test case 1 shows an error (wrong)
- ❌ Test case 2 shows and error (wrong)
- ✅ Test case 3 doesn't shows an error (fine)

**Expected result:**

All these test cases should not show any error as all this approaches are valid

### eslint-plugin-react version

v7.32.2

### eslint version

v8.50.0

### node version

v18.12.1

**Comment by ljharb** [comment]
That seems like a very strange pattern. The function should be defined at the module level, or better, don’t use a function at all and just assign to a variable - IIFEs are obsolete in a world with modules.

**Comment by avasuro** [comment]
It's a metter of taste :)

I prefer to have all my JSX code grouped together at the very bottom of the file. When I use a variable to store parts of the JSX markup, I find it less convenient because I have to constantly jump between the variable and the "main" JSX (within the return statement), which I don't like. To me, the drawback of using an IIFE (which can be less readable than normal JavaScript code) is outweighed by the convenience of not having to navigate up and down in the code.

**Comment by ljharb** [comment]
Either way, this is a bizarre coding pattern that I've never seen before and never expect to see again, so it's not worth supporting in this plugin at this time.
